Implement Adaptive Speed mod

hlysine opened the pull request:
Discussed in #17028 

This mod changes track rate according to the speed that the user is playing at. For each hit, a target rate is calculated and the most recent target rates are averaged to get the final rate change.

Notes:
- There is no way to know the accurate track rate before playing, so the mod simply reports a constant rate in `ApplyToRate`. Even though it is inaccurate, it doesn't seem to break anything in my testing.
- Storyboard samples always play at the initial rate and do not vary with the track rate. This affects `ModTimeRamp` as well. See #17053

hlysine committed:
Add adaptive speed mod

hlysine committed:
Add initial rate setting

hlysine committed:
Improve adaptive speed algorithm and add rewind support

hlysine committed:
Use smooth speed change

hlysine committed:
Allow the mod to properly react to nested hit objects

hlysine committed:
Tweak `average_count`

apollo-dw commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+            sample.AddAdjustment(AdjustableProperty.Frequency, SpeedChange);
+        }
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
```
From a difficulty calculation standpoint, this should be `rate * slowest_rate` I'm pretty sure to match worst case scenarios. From testing the mod out, it's pretty feasible to slow down the map to very low rates whilst hitting all 300s (if you hit late on every note)

hlysine replied:
Difficulty calculation uses `ApplyToTrack` instead of `ApplyToRate` for rate determination, and it's not a good idea to use `slowest_rate` in `ApplyToTrack` because song select plays at that rate too.

apollo-dw replied:
Ah, that's down to me misremembering - since I used `ApplyToRate` in `ClockWithMods` in https://github.com/ppy/osu/pull/14831. Nonetheless, I'm pretty confident in my assessment on difficulty with the current proposed score structure (unless score data somehow starts including state of adaptive mods like this one) - maybe it's too early to action upon it in this PR though.

bdach replied:
I'd say the diffcalc implications don't need to be addressed at this stage yet. Other variable rate mods are already handled on a best-effort basis. Would prefer to keep this as is for now.

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+        private ITrack track;
+        private HUDOverlay overlay;
+
+        private readonly List<double> recentRates = Enumerable.Range(0, average_count).Select(_ => 1d).ToList();
```
```suggestion
        private readonly List<double> recentRates = Enumerable.Repeat(1d, average_count).ToList();
```

Shame that C# doesn't have a baked-in deque structure but alas.

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+
+        public void ApplyToHUD(HUDOverlay overlay)
+        {
+            // this is only used to transform the SpeedChange bindable
+            this.overlay = overlay;
```
This is really bad. Why is this mod not using `IUpdatableByPlayfield` like `ModTimeRamp` even is?

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+            InitialRate.TriggerChange();
+            AdjustPitch.TriggerChange();
+            recentRates.Clear();
+            recentRates.AddRange(Enumerable.Range(0, average_count).Select(_ => InitialRate.Value));
```
```suggestion
            recentRates.AddRange(Enumerable.Repeat(InitialRate.Value, average_count));
```

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+
+        private void applyPitchAdjustment(ValueChangedEvent<bool> adjustPitchSetting)
+        {
+            // remove existing old adjustment
```
```suggestion
```

Comment brings no value.

peppy replied:
Is there a reason the removal/addition is even needed? The adjustments are bindables, so I'd hope they could be adjusted using `.Value = ` rather than `Remove` then `Add`.

bdach replied:
Note that they potentially affect different adjustment types (frequency vs tempo). It is a bit hidden though, made the same mistake when initially reading.

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+                dequeuedRates.Add(result.HitObject, recentRates[0]);
+                recentRates.RemoveAt(0);
+
+                overlay.TransformBindableTo(SpeedChange, recentRates.Average(), 100);
```
This is going to create a lot of transforms, potentially overlapping, which is going to be a performance concern. The usual way of handling that is using `Interpolation.DampContinuously()` or similar.

hlysine replied:
Thanks, that's the function I was missing. Wasn't sure about using transforms in the first place.

bdach commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,183 @@
+
+            foreach (HitObject hitObject in hitObjects)
+            {
+                double prevEndTime = endTimes.LastOrDefault(ht => !Precision.AlmostBigger(ht, hitObject.GetEndTime()));
```
This is unnecessarily quadratic. Could be improved to O(n log n) by binary searching `endTimes` instead. Several other examples of this exist in this repository.

hlysine replied:
Wow, didn't know about the bitwise complement trick in `List<T>.BinarySearch` at all. What a weird hack.

hlysine committed:
Use `Enumerable.Repeat`

hlysine committed:
Extract duplicated conditionals

hlysine committed:
Remove pointless comment

hlysine committed:
Use binary search in `ApplyToBeatmap`

hlysine committed:
Interpolate speed change using `IUpdatableByPlayfield`

peppy commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,188 @@
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
+
+        private void applyPitchAdjustment(ValueChangedEvent<bool> adjustPitchSetting)
```
Rename to `onAdjustPitchChanged` or `adjustPitchChanged`. This is still misleading to my eyes.

peppy commented on osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs:
```diff
@@ -0,0 +1,188 @@
+        // caching them here for easy access
+        private readonly Dictionary<HitObject, double> previousEndTimes = new Dictionary<HitObject, double>();
+
+        // record the value removed from recentRates when an object is hit
```
convert to xmldoc (`///` should autocomplete to an xmldoc section).

peppy reviewed (changes requested):
Not reviewing the rest of the code because it's a bit too much. Probably needs a rewrite/refactor pass by someone else before being acceptable though.

hlysine committed:
Convert comments to xmldoc

hlysine committed:
Rename `applyPitchAdjustment` to `adjustPitchChanged`

hlysine committed:
Reorder members

bdach commented:
@peppy do you mind expounding on what you'd like addressed in a refactor pass here? Difficult to tell from just the above review. This diff seemed mostly sane when I read through it.

peppy commented:
Maybe it's okay, I just found the main math portion a bit hard to mentally parse. If you're fine with it go ahead, my remarks weren't intended to be blocking.

bdach commented:
I'll say that some of the xmldocs here don't really help understanding too much (or even make it harder). I'll try rewording those and maybe it'll help some.

bdach committed:
Move and reword docs of allowable rate range constants

bdach committed:
Add extended documentation of adaptive speed mod machinations

bdach commented:
I've fired off a pair of commits in an attempt to convey what this does better - it's generally just a few renames here and there and a decent amount of comments (with an illustrative example).

@peppy please check if you find this more understandable now; @hlysine please check that I haven't mangled anything too badly in the process.

bdach committed:
Fix slightly wrong references in xmldocs

peppy commented:
Playability of this feels a bit weird. Some suggestions for adjustment:

- When missing, the rate should probably decrease a bit. It's easy to "accidentally" increase the rate while playing really badly by hitting an object near the extent of its hittable range (in the early direction) and causing a suddenly playback speed instead.
- The amount by which the speed changes should be capped to a more sane value. Right now it's possible to go from 50% to 250% in a few clicks, which feels really bad.

I'm not sure how I feel about the rate limit extents being larger than normal. Maybe okay, but not sure.

Also this mod is called "Adaptive Speed" right now, but I don't believe we refer to rate adjusts as "speed" anywhere else. May be wise to rename it in line with terminology used elsewhere?

peppy reviewed (changes requested):
As proposed.

hlysine committed:
Rename `approximatedRates` to `ratesForRewinding` and update xmldoc

hlysine commented:
> I've fired off a pair of commits in an attempt to convey what this does better - it's generally just a few renames here and there and a decent amount of comments (with an illustrative example).

Thanks for the xmldoc. They are mostly accurate and explain the mod quite well, but the explanation for `approximatedRates` is a bit off. I have renamed it to `ratesForRewinding` and changed the xmldoc to hopefully clear things up. I've also re-ordered some statements near its usage to group the related logic together.

hlysine committed:
Cap speed change per hit and apply a speed decrease on miss

hlysine commented:
> When missing, the rate should probably decrease a bit. It's easy to "accidentally" increase the rate while playing really badly by hitting an object near the extent of its hittable range (in the early direction) and causing a suddenly playback speed instead.

Implemented. Track rate decreases by 5% for each miss.

> The amount by which the speed changes should be capped to a more sane value. Right now it's possible to go from 50% to 250% in a few clicks, which feels really bad.

The relative rate change caused by a single hit is now clamped between 0.8 and 1.25. This is just here to prevent a huge rate change caused by a single mistimed hit. If the player intends to go from 50% to 250%, I don't see it as a problem and it should still be possible in a few clicks.

> I'm not sure how I feel about the rate limit extents being larger than normal. Maybe okay, but not sure.

I think 250% is fine, but I'll admit that 40% is pretty unplayable. That being said, I feel like we shouldn't allow players to start at 50% and stay there by consistently hitting notes late. That's why the extended limits are there.

> Also this mod is called "Adaptive Speed" right now, but I don't believe we refer to rate adjusts as "speed" anywhere else. May be wise to rename it in line with terminology used elsewhere?

I see "speed" as being interchangeable with "rate". The setting in Double Time is called "Speed increase", for example. Also, "Adaptive Speed" just sounds better than "Adaptive Rate" to me.

peppy commented:
I dunno, still feels like this requires a hell of a lot more tuning to me. I can't play because it adjusts in a direction I don't expect it to.

Will leave to others to make a call on it, I'd recommend play testing before considering merging.

hlysine committed:
Merge branch 'master' into mod-adaptive-speed

hlysine committed:
Scale rate adjustments based on hit timing consistency and tweak some related numbers

hlysine commented:
After a lot of playtesting, I feel like the mod was being too harsh on the accuracy requirement and was distorting the song too much, which was probably why you find the mod adjusting in an unexpected direction. 

I added some more logic during rate adjustment to scale the adjustment based on the player's timing consistency. Large changes to track rate are only made if the player consistently hits their notes too fast/too slow. I also tweaked some of the constants to make the mod slightly easier in general.

With this new change, you should be able to get through a song like normal as long as you have acceptable accuracy (something above 90%, for example). But of course, it is still possible to go from a 50% to 250% rate if you wish, since that's the main point of the mod.

respektive commented:
This does feel a lot better than before now, especially on non-streaming parts it isn't moving back and forth nearly as much as before anymore.

peppy commented:
I think this is feeling pretty good now.

smoogipoo commented:
Any reason this isn't enabled for the catch ruleset?

peppy commented:
osu!catch hits are never early or late, so the only effect this mod would have is slowing the gameplay down on missing.

smoogipoo commented:
Yeah intuitively that's how I think it'd work, but seems like it doesn't.

peppy commented:
If it was to be implemented for catch i think you'd need to have the speed revert to 100% for each successful hit, but never go above. Or something like that.

smoogipoo reviewed (approved):
Very rudimentary pass over this, :+1:

hlysine commented:
> Yeah intuitively that's how I think it'd work, but seems like it doesn't.

It doesn't because catch objects all have `EmptyHitWindows`, and this mod ignores all such hit objects.

> If it was to be implemented for catch i think you'd need to have the speed revert to 100% for each successful hit, but never go above. Or something like that.

Sounds good, but it also sounds like a separate mod to me. I don't play catch at all so I'm not sure.
